bittensor/utils/version.py

```diff
@@ -0,0 +1,74 @@
+from typing import Optional
+from pathlib import Path
+import time
+from packaging.version import Version
+
+import bittensor
+import requests
+
+VERSION_CHECK_THRESHOLD = 86400
+
+
+def _get_version_file_path() -> Path:
+    return Path.home() / ".bittensor" / ".last_known_version"
+
+
+def _get_version_from_file(version_file: Path) -> Optional[str]:
+    if not version_file.exists():
+        bittensor.logging.debug("No bitensor version file found")
+        return None
+
+    mtime = version_file.stat().st_mtime
+    bittensor.logging.debug(f"Found version file, last modified: {mtime}")
+    diff = time.time() - mtime
+
+    if diff >= VERSION_CHECK_THRESHOLD:
+        bittensor.logging.debug("Version file expired")
+        return None
+
+    try:
+        return version_file.read_text()
+    except OSError:
+        bittensor.logging.exception("Failed to read version file")
+        return None
+
+
+def _get_version_from_pypi(timeout: int = 15) -> str:
+    bittensor.logging.debug(
+        f"Checking latest Bittensor version at: {bittensor.__pipaddress__}"
+    )
+    try:
+        response = requests.get(bittensor.__pipaddress__, timeout=timeout)
+        latest_version = response.json()["info"]["version"]
+        return latest_version
+    except requests.exceptions.RequestException:
+        bittensor.logging.exception("Failed to get latest version from pypi")
+        raise
+
+
+def get_and_save_latest_version(timeout: int = 15) -> str:
+    version_file = _get_version_file_path()
+
+    if last_known_version := _get_version_from_file(version_file):
+        return last_known_version
+
+    latest_version = _get_version_from_pypi(timeout)
```

Comment:
reminder: bitensor.cli is not only user of `version_checking` function

Please put the old try-except into back into `version_checking` function

Reply:
Here is the previous implementation of `version_checking`:

```python
    try:
        bittensor.logging.debug(
            f"Checking latest Bittensor version at: {bittensor.__pipaddress__}"
        )
        response = requests.get(bittensor.__pipaddress__, timeout=timeout)
        latest_version = response.json()["info"]["version"]
        version_split = latest_version.split(".")
        latest_version_as_int = (
            (100 * int(version_split[0]))
            + (10 * int(version_split[1]))
            + (1 * int(version_split[2]))
        )

        if latest_version_as_int > bittensor.__version_as_int__:
            print(
                "\u001b[33mBittensor Version: Current {}/Latest {}\nPlease update to the latest version at your earliest convenience. "
                "Run the following command to upgrade:\n\n\u001b[0mpython -m pip install --upgrade bittensor".format(
                    bittensor.__version__, latest_version
                )
            )

    except requests.exceptions.Timeout:
        bittensor.logging.error("Version check failed due to timeout")
    except requests.exceptions.RequestException as e:
        bittensor.logging.error(f"Version check failed due to request failure: {e}")
```

And here is the calling block of code in `cli.py`:
```python
        try:
            bittensor.utils.version_checking()
        except:
            # If version checking fails, inform user with an exception.
            raise RuntimeError(
                "To avoid internet-based version checking, pass --no_version_checking while running the CLI."
            )
```

So, it seems that the latter except block wasn't reachable at all, we would just log the error inside the function instead. Do you think that function was meant to fail silently in the first place?

Reply:
This argument is irrelevant seeing how cli.py is not the only place where this is called from

Just some random search shows people are not always putting it in try except: https://github.com/Agent-Artificial/commune/blob/75e01a735dbca3857c6be0accb5c436b04e38e04/commune/vali/bt/validator.py#L788


If someone wants to do custom messages etc, they can call `get_and_save_latest_version` themselves.

Reply:
I think the argument is still relevant, cause we have a block of code in `cli.py` that is effectively unreachable. This cannot be simply solved due to the constraint you mentioned. I ended up "soft deprecating" the `version_checking` function and introducing a new `check_version` function, where the raised exception is documented in the docstring.